wa/workloads/speedometer/__init__.py

```diff
@@ -12,57 +12,295 @@
+
+        file_to_modify = "/data/data/{}/shared_prefs/{}".format(
+            self.chrome_package, self.preferences_xml
+        )
+
+        self.target.pull(file_to_modify, context.output_directory, as_root=True)
+
+        with open(
+            os.path.join(context.output_directory, self.preferences_xml)
+        ) as read_fh:
+            lines = read_fh.readlines()
+
+            # Add additional elements for the preferences XML to the
+            # _second-last_ line
+            for line in [
+                '<boolean name="first_run_flow" value="true" />\n',
+                '<boolean name="first_run_tos_accepted" value="true" />\n',
+                '<boolean name="first_run_signin_complete" value="true" />\n',
+            ]:
+                lines.insert(len(lines) - 1, line)
+
+            with open(
+                os.path.join(context.output_directory, self.preferences_xml + ".new"),
+                "w",
+            ) as write_fh:
+                for line in lines:
+                    write_fh.write(line)
+
+        # Make sure ownership of the original file is preserved.
+        user_owner, group_owner = self.target.execute(
+            "ls -l {}".format(file_to_modify),
+            as_root=True,
+        ).split()[2:4]
+
+        self.target.push(
+            os.path.join(context.output_directory, self.preferences_xml + ".new"),
+            file_to_modify,
+            as_root=True,
+        )
+
+        self.target.execute(
+            "chown {}.{} {}".format(user_owner, group_owner, file_to_modify),
+            as_root=True,
+        )
+
+    def run(self, context):
+        super(Speedometer, self).run(context)
+
+        browser_launch_cmd = "am start -a android.intent.action.VIEW -d {} {}".format(
+            self.speedometer_url, self.chrome_package
+        )
+        self.target.execute(browser_launch_cmd)
+
+        # Wait 60 seconds before we start polling the system to check which
+        # processes are busy.
+        time.sleep(60)
+
+        # We can only consider the benchmark complete and proceed if we see 5
+        # times _in a row_ that a process containing the string
+        # 'sandboxed_process' is not the hottest process in the system. We will
+        # sleep for 2 seconds between every check.
+        top_cmd = "{} top -b -n1".format(self.target.busybox)
+
+        def _top_output_for_busiest_process():
+            # busybox's top always outputs the Mem/CPU/Load/Header lines first,
+            # so line 5 is the hottest process.
+            return self.target.execute(top_cmd).split("\n")[4]
+
+        def _get_all_processes():
+            return self.target.execute(top_cmd).split("\n")[4:]
+
+        search_term = self.chrome_package + ":sandbox"
+        if self.render_process_search_term is not None:
+            search_term = self.render_process_search_term
+        render_processes = [x for x in _get_all_processes() if search_term in x]
+        if len(render_processes) == 0:
```

Comment:
pylint nit: "Do not use `len(SEQUENCE)` to determine if a sequence is empty (len-as-condition)"

Reply:
I think this is actually a bug in the version of pylint you're using. It is designed to catch use of `len(X)` as the sole condition in an if statement, when `len(X) == 0` is perfectly OK, and in my opinion, more readable as its intent is more explicit. This particular version of pylint (assuming you're using 1.9.2) rejects this line, but later versions do not - I didn't see this initially when I installed the latest version of pylint through pip (2.5.3). See https://github.com/PyCQA/pylint/issues/2684 for the issue which led to pylint being fixed in light of this.

Reply:
Hmm I see. Yes you're correct, we're still using v1.9.2 as at the time we hit a bug in the later version but we could probably do with an upgrade by now.